File: medium/1396_design_underground_system/solution.py
class UndergroundSystem:
    def __init__(self):
        self.system: dict[int, tuple[str, int]] = {}
        self.travels: dict[tuple[str, str], tuple[int, int]] = {}

    def checkIn(self, id: int, stationName: str, t: int) -> None:
        self.system[id] = (stationName, t)

    def checkOut(self, id: int, stationName: str, t: int) -> None:
        startStation, startTime = self.system[id]

        sum, count = self.travels.get((startStation, stationName), (0, 0))
        sum += t - startTime
        count += 1
        self.travels[(startStation, stationName)] = (sum, count)

        self.system.pop(id)

    def getAverageTime(self, startStation: str, endStation: str) -> float:
        sum, count = self.travels[(startStation, endStation)]
        return sum / count

File: medium/1396_design_underground_system/solution.cpp
#include <string>
#include <unordered_map>

class UndergroundSystem {
public:
  UndergroundSystem() {
    system = std::unordered_map<int, std::pair<std::string, int>>();
    travels = std::unordered_map<
        std::string, std::unordered_map<std::string, std::pair<int, int>>>();
  }

  void checkIn(int id, std::string stationName, int t) {
    system[id] = {stationName, t};
  }

  void checkOut(int id, std::string stationName, int t) {
    const auto &[startStation, startTime] = system[id];

    auto &[sum, count] = travels[startStation][stationName];
    sum += t - startTime;
    ++count;

    system.erase(id);
  }

  double getAverageTime(std::string startStation, std::string endStation) {
    const auto &[sum, count] = travels[startStation][endStation];
    return static_cast<double>(sum) / static_cast<double>(count);
  }

private:
  std::unordered_map<int, std::pair<std::string, int>> system;
  std::unordered_map<std::string,
                     std::unordered_map<std::string, std::pair<int, int>>>
      travels;
};
